### PR TITLE
SortByKey and GroupByKey: Invalid input.

### PR DESCRIPTION
**Pull request by aosyatnik**
### Purpose

Link to YouTrack:
[MAGN-8753](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8753) Add 'Invalid Input' error message to SortByKey and GroupByKey

I thought that would be the easiest task. I was wrong.

I created SortByKey and GroupByKey as UI nodes, they were called from FunctionObject.ds file. But to show any warning we need throw exception. In design script they is no way of creating exception.

So, I had to make SortByKey and GroupByKey as zero touch nodes. Also I duplicated old tests, but rewrote them as for zero touch node tests.

### Declarations

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@mjkkirschner 

### FYIs

@Racel

### CODE REVIEW

**Comment by aosyatnik** [comment]
All tests passed.
image

File: src/Libraries/CoreNodes/List.cs [hunk 1]
```diff
@@ -235,6 +236,87 @@ public static IDictionary Deconstruct(IList list)
...
+                        numberPair => numberPair.item == pair.item && numberPair.key == pair.key));
+
+            // Sort.
+            numberKeyPairs = numberKeyPairs.OrderBy(pair => Convert.ToDouble(pair.key));
+            keyPairs = keyPairs.OrderBy(pair => pair.key);
+
+            // First items with number keys, then items with letter keys.
+            var sortedPairs = numberKeyPairs.Concat(keyPairs);
+
+            var sortedList = sortedPairs.Select(x => x.item);
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/List.cs]: I think it might be best to convert ToList() at this point before returning, this is done with the group by nodes and other list nodes return lists not ienums.

File: src/Libraries/CoreNodes/List.cs [hunk 2]
```diff
@@ -235,6 +236,87 @@ public static IDictionary Deconstruct(IList list)
...
+                return null;
+
+            var containsSublists = keys.Cast<object>().Any(key => key is IList || key is ICollection);
+            if (containsSublists)
+            {
+                throw new ArgumentException(Resources.InvalidKeysErrorMessage);
+            }
+
+            if (list.Count != keys.Count)
+                return null;
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodes/List.cs]: it's been so many iterations I have forgotten :), is there a reason why we just return null here and don't have a warning message explaining why?

File: test/Libraries/CoreNodesTests/ListTests.cs [hunk 3]
```diff
@@ -950,5 +950,206 @@ public static void Chop6()
 
             Assert.Throws<InvalidCastException>(() => { List.Chop(list, lengths); });
         }
+
+        [Test]
```
Comment (mjkkirschner) [inline test/Libraries/CoreNodesTests/ListTests.cs]: are there any tests for SortByFunction? - are these covered somewhere else in the tests? If not can we add some?

Comment (aosyatnik) [inline test/Libraries/CoreNodesTests/ListTests.cs]: I added additional test for GrouByFunction. Test for SortByFunction was created before me...

File: src/Libraries/CoreNodes/List.cs (hunk 2 above)
Comment (aosyatnik) [inline src/Libraries/CoreNodes/List.cs]: There wasn't any special reason why we didn't show warning. The only reason why we didn't do it, because it was design script and in design there is no opportunity to generate warning. At this place I can throw exception with some clear description. 
@Racel , would you like to show warning, if number of list items doesn't equal number of key items? If you do want, can you provide appropriate message?
Thanks.

**Comment by Racel** [comment]
@aosyatnik - when # of items does not equal # of keys, please provide the warning: "Number of items does not match the number of keys"

**Comment by Racel** [comment]
@aosyatnik - also, see my comment in Youtrack. Can we change the wording of the invalid input warning? I have added the new text to the description in Youtrack.

**Comment by aosyatnik** [comment]
@mjkkirschner , I updated PR, could you take a look?

File: src/Libraries/CoreNodeModels/Properties/Resources.resx [hunk 4]
```diff
@@ -584,40 +584,4 @@
   <data name="SelectionPointsOutputPortName" xml:space="preserve">
     <value>Points</value>
   </data>
-  <data name="ListGroupByKeyDescription" xml:space="preserve">
-    <value>Group items into sub-lists based on their like key values</value>
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/Properties/Resources.resx]: sorry I missed this the first time around - is there a reason you have not added all of these descriptions and tooltips to the zero touch nodes?

File: src/Engine/ProtoCore/ExtendedLibraries/FunctionObject.ds [hunk 5]
```diff
@@ -479,33 +479,7 @@ def SortByFunction(list : var[]..[], func:_FunctionObject)
         {
             _keys[_index] = __Apply(func, list[_index]);
         }
-        return = Sorting.sortByKey(list, _keys)[0];
```
Comment (mjkkirschner) [inline src/Engine/ProtoCore/ExtendedLibraries/FunctionObject.ds]: another concern I have is this file:
https://github.com/DynamoDS/Dynamo/blob/11f90d72e1047fcea9ece3818e90cf420890864d/src/DynamoCore/BuiltInAndOperators/BuiltIn.Migrations.xml

just a question for you: does this migration work correctly now that sortbykey and group by key are zero touch nodes?

**Comment by mjkkirschner** [comment]
@aosyatnik this looks really good, thanks for your work!

Theres one more issue with the zero touch node losing the nice strings we had to tooltips and descriptions from the other nodes, and just a concern about the migration file that I would like you to check out. Other than that looks good to merge.

**Comment by aosyatnik** [comment]
@mjkkirschner migration for BuiltIn nodes works properly. I added migration for UI nodes (SortByKey and GroupByKey).
We don't lose descriptions. I added them as xml comment <summary>.
image 
image

**Comment by aosyatnik** [comment]
Passed. 
image
